source/extensions/filters/http/oauth2/filter.h

```diff
@@ -0,0 +1,232 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "envoy/common/matchers.h"
+#include "envoy/config/core/v3/http_uri.pb.h"
+#include "envoy/extensions/filters/http/oauth2/v3alpha/oauth.pb.h"
+#include "envoy/http/header_map.h"
+#include "envoy/server/filter_config.h"
+#include "envoy/stats/stats_macros.h"
+#include "envoy/stream_info/stream_info.h"
+#include "envoy/upstream/cluster_manager.h"
+
+#include "common/common/assert.h"
+#include "common/common/matchers.h"
+#include "common/config/datasource.h"
+#include "common/formatter/substitution_formatter.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/header_utility.h"
+#include "common/http/rest_api_fetcher.h"
+#include "common/http/utility.h"
+
+#include "extensions/filters/http/common/pass_through_filter.h"
+#include "extensions/filters/http/oauth2/filter.h"
+#include "extensions/filters/http/oauth2/oauth.h"
+#include "extensions/filters/http/oauth2/oauth_client.h"
+
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Oauth2 {
+
+class OAuth2Client;
+
+// Helper class used to fetch secrets (usually from SDS).
+class SecretReader {
+public:
+  virtual ~SecretReader() = default;
+  virtual std::string clientSecret() const PURE;
+  virtual std::string tokenSecret() const PURE;
```

Comment:
make this return a `const std::vector<uint8_t>&`? Would avoid copy / conversion below.